### ISSUE TITLE
Assert failure after S.stop

### ISSUE THREAD

**Issue by art-w** [dbuenzli/react#18]
The following code exhibits a small effect leak:

``` ocaml
let x, set_x = S.create 0

let y =
  S.bind x (fun v ->
    let id = Random.int 10000 in
    let s = S.map (fun _ -> Printf.printf "hi %i\n%!" id) x in
    (* S.stop s ; *)
    s)

let () =
  List.iter
    (fun v -> Printf.printf "---\n%!" ; Gc.full_major () ; set_x v)
    [1;2]
```

Outputs:

```
hi 9344

---
hi 9344   <- leak from previous step
hi 6685

---
hi 6685   <- here too
hi 182
```

If we try to naively stop the inner signal early, we get:

```
hi 9344

---
Fatal error: exception Assert_failure("src/react.ml", 422, 51)
```

The documentation states that `S.stop` may not take effect immediately if executed during an update step: But here, `S.stop` _does_ kill the signal early, which puts `S.bind` in trouble. I would expect the following output:

```
hi 9344

---
hi 6685

---
hi 182
```

If that's not possible, it would be nice to raise a proper exception and document when it is safe to call `S.stop`.

**Comment by dbuenzli** [comment]
> But here, S.stop does kill the signal early, which puts S.bind in trouble. I would expect the following output:
> 
> hi 9344
> ---
> hi 6685
> ---
> hi 182

The documentation of `S.stop` says it may or it may not update in the step so you cannot expect this output. In this case you create a signal and stop it in the same update step, remember that an update step is conceptually instantaneous, so I'd say that you are doing something fishy semantically. 

In any case `S.stop` should not raise. So this is a a bug.

**Comment by dbuenzli** [comment]
> The documentation states that S.stop may not take effect immediately if executed during an update step

Note, this not exactly what the doc says (altough I cannot say that the wording is exactly enlighting). The doc says that if S.stop is executed in a step on a signal the signal may still update in the step (in the sense take the value it should have if it wasn't stopped). In fact S.stop always happens immediately, however when you apply it on a signal in a step I cannot guarantee that this signal didn't already update in the step.

**Comment by art-w** [comment]
Yes, I understand how S.stop works and what it implies. You're right that any subset of the "hi ..." would also be an acceptable behaviour; and that this code is using React the wrong way.

I believe S.bind is the one that raises an assertion, since it tries to switch to an uninitialized (and stopped) signal.

**Comment by dbuenzli** [comment]
> I believe S.bind is the one that raises an assertion, since it tries to switch to an uninitialized (and stopped) signal.

Mmmh this tend to indicate that the fix will be to add to the documentation do not use `S.stop` in an update step...

**Comment by dbuenzli** [comment]
> Mmmh this tend to indicate that the fix will be to add to the documentation do not use S.stop in an update step...

So looking at the implementation of `S.stop` suggested that we could simply look if the signal is not initialized in which case we know we are in an update step and will be scheduled for update, we let the system do that so that everything stabilizes correctly and stop the signal at the end of the update cycle via an end of step operation.

This is implemented in 3f6ee420f4da5516ed and we get exactly the output you expected.
